src/main/perl/Download.pm

```diff
@@ -0,0 +1,142 @@
+# ${license-info}
+# ${developer-info}
+# ${author-info}
+# ${build-info}
+
+package CAF::Download;
+
+use strict;
+use warnings;
+
+# For re-export only
+use CAF::Download::URL qw(set_url_defaults);
+use parent qw(CAF::ObjectText Exporter CAF::Download::URL CAF::Download::Retrieve);
+
+use Readonly;
+use LC::Exception qw (SUCCESS);
+
+our @EXPORT_OK = qw(set_url_defaults);
+
+# TODO: dependencies on curl and kinit
+
+Readonly::Hash my %DOWNLOAD_METHODS => {
+    http => [qw(lwp curl)], # try https, if not, try http
+    https => [qw(lwp curl)], # only https
+    file => [qw(lwp)],
+};
+
+Readonly::Array my @DOWNLOAD_PROTOCOLS => sort keys %DOWNLOAD_METHODS;
+
+# TODO: can we mix and match x509/krb5 also for security like TLS?
+# The GSSAPI doesn't require TLS, it has encryption
+# gssapi here means use the perl GSSAPI bindings to generate the tokens etc
+# kinit means to use commandline tools like kinit/kdestroy
+# x509/lwp means have LWP handle X509 (TLS + X509 auth)
+# TODO: does kinit imply GSSAPI usage?
+Readonly::Hash my %DOWNLOAD_AUTHENTICATION => {
+    krb5 => [qw(gssapi kinit)],
+    x509 => [qw(lwp)],
+};
+
+# Disclaimer: inspired by
+#    NCM::Component::download (15.8)
+#    EDG::WP4::CCM::Fetch (15.8)
+#    File::Fetch (0.48)
+
+=pod
+
+=head1 NAME
+
+CAF::Download - Class for downloading content from remote servers.
+
+=head1 SYNOPSIS
+
+    use CAF::Download;
+
+    my $dl = CAF::Download->new(['https://somewhere/myfile']);
+    print "$dl"; # stringification
```

Comment:
i'll address error handling in the `ObjectText` class (but i have to admit this really is not trivial in `CAF`, i think i've only implemented this once myself).
it would actually be nice to know what is expected. is it ok for a `CAF` class to throw exceptions? most of current code do not handle exceptions being thrown